src/datastore/postgres-store.ts

```diff
@@ -5644,6 +5644,38 @@ export class PgDataStore
         `,
         [address, maxBlockHeight]
       );
+
+      if (query.rowCount > 0) {
+        const rows: { name: string; block_height: number }[] = [];
+        for (const row of query.rows) {
+          const query_latest = await client.query<{ name: string }>(
+            `
+            SELECT name FROM (
+              (
+                SELECT name
+                FROM names
+                WHERE name = $1
+                AND registered_at > $2
+                AND canonical = true AND microblock_canonical = true
+              )
+              UNION ALL (
+                SELECT fully_qualified_subdomain as name
+                FROM subdomains
+                WHERE fully_qualified_subdomain = $1
+                AND block_height > $2
+                AND canonical = true AND microblock_canonical = true
+              )
+            ) results
+            `,
+            [row.name, row.block_height]
+          );
+
+          if (query_latest.rowCount === 0) rows.push(row); //if there is no names exist at greater height
```

Comment:
Instead of moving results back and forth from postgres to node to filter, you could try doing all the filtering in a single query. The idea would be:

1. In a common table expression or subquery, get all names ever owned by $1
2. From step 1 name results, query to get the latest owner address (limit 1 ordered by block desc)
3. From step 2 results, select only the names where the owner is $1

Reply:
This way you avoid the O(n) problem in node and leave that to the DB to handle